### PR TITLE
reflection: allow `names` to return `using`-ed names

### LINKED ISSUE #36529: Visibility of `using`ed names in a module
```julia
julia> module A
       module B
       f(x) = 1
       end
       using .B: f
       end
Main.A

julia> names(A; all=true, imported=true)
6-element Array{Symbol,1}:
 Symbol("#eval")
 Symbol("#include")
 :A
 :B
 :eval
 :include
```

Is there a way to discover ~what `f` refers to inside `A`~ that `A` also has a meaning for `f`? If you change `using .B: f` to `import .B: f`, then it's visible.

This appears to be related to a significant performance issue for SymbolServer (which is used by vscode's languageserver indexing machinery), see https://github.com/julia-vscode/SymbolServer.jl/pull/177.

### PR DESCRIPTION
**Pull request by aviatesk**
This commit makes it possible for `names` to return `using`-ed names as well:
```julia
julia> using Base: @assume_effects

julia> Symbol("@assume_effects") in names(@__MODULE__; usings=true)
true
```

Currently, to find all names available in a module `A`, the following steps are needed:
1. Use `names(A; all=true, imported=true)` to get the names defined by `A` and the names explicitly `import`ed by `A`.
2. Use `jl_module_usings(A)` to get the list of modules `A` has `using`-ed and then use `names()` to get the names `export`ed by those modules.

This method is implemented in e.g. REPL completions, but it has a problem: it could not get the names explicitly `using`-ed by `using B: ...` (#36529, #40356, JuliaDebug/Infiltrator.jl#106, etc.).

This commit adds a new keyword argument `usings::Bool=false` to `names(A; ...)`, which, when `usings=true` is specified, returns all names introduced by `using` in `A`.
In other words, `usings=true` not only returns explicitly `using`-ed names but also incorporates step 2 above into the implementation of `names`.

By using this new option, we can now use
`names(A; all=true, imported=true, usings=true)` to know all names available in `A`, without implementing the two-fold steps on application side.
As example application, this new feature will be used to simplify and enhance the implementation of REPL completions.

- fixes #36529

### CODE REVIEW

File: src/module.c [hunk 1]
```diff
@@ -1003,16 +1012,39 @@ JL_DLLEXPORT jl_value_t *jl_module_names(jl_module_t *m, int all, int imported)
         jl_sym_t *asname = b->globalref->name;
         int hidden = jl_symbol_name(asname)[0]=='#';
         int main_public = (m == jl_main_module && !(asname == jl_eval_sym || asname == jl_include_sym));
-        if ((b->publicp ||
+        if ((jl_atomic_load_relaxed(&b->value) != (jl_value_t*)m) &&
```
Comment (JeffBezanson) [inline src/module.c:1015]: I think this should be based on the name instead of the value?

File: src/module.c [hunk 2]
```diff
@@ -1003,16 +1012,39 @@ JL_DLLEXPORT jl_value_t *jl_module_names(jl_module_t *m, int all, int imported)
...
+    if (usings) {
+        // If `usings` is specified, traverse the list of `using`-ed modules and incorporate
+        // the names exported by those modules into the list.
+        for(int i=(int)m->usings.len-1; i >= 0; --i) {
+            jl_module_t *usinged = module_usings_getidx(m, i);
+            _jl_module_names_into_array(a, usinged, 0, 0, 0, /*exported*/1);
+            // Add the name of `usinged` itself, unless the user requested `all=true` and it's
+            // a submodule of `m`, since then its name would have already been added by
+            // `all=true`, since it's a binding in `m`.
+            if (!all || usinged->parent != m) {
```
Comment (JeffBezanson) [inline src/module.c:1043]: I'm not sure this is enough to avoid duplicates. For example the name might also be exported. We could also de-duplicate in the caller though.

File: src/module.c (hunk 1 above)
Comment (aviatesk) [reply src/module.c:1015]: It seems that this code was added in #42092, but I think it can be removed without any issues, so I plan to do that.

File: src/module.c (hunk 2 above)
Comment (aviatesk) [reply src/module.c:1054]: Yeah... I've been thinking about how to eliminate the duplication, and it seems it might be better to start by organizing the interface of `names`. With the current design and implementation, `all`/`imported` and `usings` are not necessarily orthogonal, so the duplication is inherent in the definitions.
By defining `names` as `names(x::Module; non_public::Bool=false, generated::Bool=false, imported::Bool=false, usings::Bool=false)` and refactoring the code, I think we can make the handling of such duplications more clear.

Comment (aviatesk) [reply src/module.c:1054]: The refactoring mentioned above involves more changes, so..., for now in this PR, I want to either 1) have `names` return a `Set` to eliminate duplicates on the `names` side (technically this is a breaking change?), or 2) add a warning about the possibility of duplicates in the `names` docstring. Which do you think is better?

Comment (aviatesk) [reply src/module.c:1054]: I've decided to go with the second option. Regarding the initial comment, I've found that this code itself is not necessary, so I will just remove it. The duplication issue remains (such as when there is a name conflicting with `using`-ed one), but at least there is a warning, so it shouldn't be a big problem.

**Comment by aviatesk** [comment]
Further refactoring might be needed for `names`, but this PR should be fine in the sense of adding new functionality. I plan to merge it once the tests pass without issues.

File: base/reflection.jl [hunk 3]
```diff
@@ -77,28 +77,33 @@ function fullname(m::Module)
...
-    names(x::Module; all::Bool = false, imported::Bool = false)
+    names(x::Module; all::Bool=false, imported::Bool=false, usings::Bool=false) -> Vector{Symbol}
 
 Get a vector of the public names of a `Module`, excluding deprecated names.
 If `all` is true, then the list also includes non-public names defined in the module,
 deprecated names, and compiler-generated names.
 If `imported` is true, then names explicitly imported from other modules
-are also included. Names are returned in sorted order.
+are also included.
+If `usings` is true, then names explicitly imported via `using` are also included.
```
Comment (JeffBezanson) [inline base/reflection.jl:87]: Also implicitly?

Comment (aviatesk) [reply base/reflection.jl:87]: Thanks for pointing that out. I plan to address this issue as well in #54659.

**Comment by lassepe** [comment]
It seems a bit confusing that `names(A; all=true)` does not return all names. But I guess that cannot be fixed in a non-breaking way?

**Comment by aviatesk** [comment]
Yeah it would be a breaking change.
